pkg/vmprovider/providers/vsphere2/vmlifecycle/update_status.go

```diff
@@ -164,19 +160,39 @@ func getGuestNetworkStatus(
 		}
 	}
 
-	for i := range guestInfo.Net {
-		// Skip pseudo devices.
-		if guestInfo.Net[i].DeviceConfigId != -1 {
-			status.Interfaces = append(status.Interfaces, guestNicInfoToInterfaceStatus(i, &guestInfo.Net[i]))
+	if len(guestInfo.Net) > 0 {
+		var networkInterfaces []vmopv1.VirtualMachineNetworkInterfaceSpec
+		if vmCtx.VM.Spec.Network != nil {
+			networkInterfaces = vmCtx.VM.Spec.Network.Interfaces
 		}
-	}
 
-	// Hack: the exceedingly common case is just one nic - our boostrap effectively only works with one -
-	// so do the best effort here and apply the name in that common case. We have a long ways to go until
-	// we can always line up interfaces in the Spec with what both is observed in the VM hardware config
-	// and what the GuestNicInfo provides.
-	if len(networkInterfaces) == 1 && len(status.Interfaces) == 1 {
-		status.Interfaces[0].Name = networkInterfaces[0].Name
+		slices.SortFunc(guestInfo.Net, func(a, b types.GuestNicInfo) int {
+			// Sort by the DeviceKey (DeviceConfigId) to order the guest info list by the
+			// order in the initial ConfigSpec which is the order of the []networkInterfaces.
+			// since it is immutable.
+			return int(a.DeviceConfigId - b.DeviceConfigId)
+		})
+
+		networkInterfaceIdx := 0
+		for i := range guestInfo.Net {
+			deviceKey := guestInfo.Net[i].DeviceConfigId
+
+			// Skip pseudo devices.
+			if deviceKey < 0 {
+				continue
+			}
+
+			var name string
+			if networkInterfaceIdx < len(networkInterfaces) {
+				name = networkInterfaces[networkInterfaceIdx].Name
+				networkInterfaceIdx++
+			} else {
+				// Provide a default name that can otherwise hopefully help identify the interface.
+				name = fmt.Sprintf("eth-deviceKey:%d", deviceKey)
```

Comment:
Why not just omit the name altogether? It should not be required in the status anyway.

Reply:
The name is the map key so it is required: https://github.com/vmware-tanzu/vm-operator/blob/main/api/v1alpha2/virtualmachine_network_types.go#L466

If we want to change the API, then sure we can leave it blank but TBH I'm not a huge fan of that. If nothing else, using some fallback name based on the device ID here helps track changes to the IP config pushed down from VM Tools for the dev w/o having to use like the MAC.

Reply:
If we want to include the device key, should we just introduce it as a field in the status?

Reply:
Done - remove Name as the list key and added a device key field